test/e2e/e2e_suite_test.go

```diff
@@ -401,3 +407,95 @@ func tearDown(bootstrapClusterProvider bootstrap.ClusterProvider, bootstrapClust
 		bootstrapClusterProvider.Dispose(context.TODO())
 	}
 }
+
+// resolveKubernetesVersions looks at Kubernetes versions set as variables in the e2e config and sets them to a valid k8s version
+// that has an existing capi offer image available. For example, if the version is "stable-1.22", the function will set it to the latest 1.22 version that has a published reference image.
+func resolveKubernetesVersions(config *clusterctl.E2EConfig) {
+	skus := getAllCAPIImageSkus(context.TODO(), os.Getenv(AzureLocation))
+	versions := parseImageSkuNames(skus)
+
+	if config.HasVariable(capi_e2e.KubernetesVersion) {
+		resolveKubernetesVersion(config, versions, capi_e2e.KubernetesVersion)
+	}
+	if config.HasVariable(capi_e2e.KubernetesVersionUpgradeFrom) {
+		resolveKubernetesVersion(config, versions, capi_e2e.KubernetesVersionUpgradeFrom)
+	}
+	if config.HasVariable(capi_e2e.KubernetesVersionUpgradeTo) {
+		resolveKubernetesVersion(config, versions, capi_e2e.KubernetesVersionUpgradeTo)
+	}
+}
+
+func resolveKubernetesVersion(config *clusterctl.E2EConfig, versions semver.Versions, varName string) {
+	v := getLatestSkuForMinor(config.GetVariable(varName), versions)
+	if _, ok := os.LookupEnv(varName); ok {
+		Expect(os.Setenv(varName, v)).To(Succeed())
+	}
+	config.Variables[varName] = v
+}
+
+// getAllCAPIImageSkus returns all skus for the capi offer under the "cncf-upstream" publisher.
+func getAllCAPIImageSkus(ctx context.Context, location string) []string {
+	settings, err := auth.GetSettingsFromEnvironment()
+	Expect(err).NotTo(HaveOccurred())
+	subscriptionID := settings.GetSubscriptionID()
+	authorizer, err := settings.GetAuthorizer()
+	Expect(err).NotTo(HaveOccurred())
+	imagesClient := compute.NewVirtualMachineImagesClient(subscriptionID)
+	imagesClient.Authorizer = authorizer
+
+	Byf("Finding image skus for offer %s/%s in %s", capiImagePublisher, capiOfferName, location)
+
+	res, err := imagesClient.ListSkus(ctx, location, capiImagePublisher, capiOfferName)
+	Expect(err).NotTo(HaveOccurred())
+
+	var skus []string
+	if res.Value != nil {
+		skus = make([]string, len(*res.Value))
+		for i, sku := range *res.Value {
+			// we have to do this to make sure the SKU has existing images
+			// see https://github.com/Azure/azure-cli/issues/20115.
+			res, err := imagesClient.List(ctx, location, capiImagePublisher, capiOfferName, *sku.Name, "", nil, "")
+			Expect(err).NotTo(HaveOccurred())
+			if res.Value != nil && len(*res.Value) > 0 {
+				skus[i] = *sku.Name
+			}
+		}
+	}
+	return skus
+}
+
+// parseImageSkuNames parses SKU names in format "k8s-1dot17dot2-ubuntu-1804" to extract the Kubernetes version.
+// it returns a sorted list of all k8s versions found.
+func parseImageSkuNames(skus []string) semver.Versions {
```

Comment:
Is it worth turning this into a proper validation func (that returns an error if one of the input strings is malformed), and are there other places in the codebase where we process image strings that would benefit from add'l validation?